quantum/util/quantum_sequencer_lite.h

```diff
@@ -0,0 +1,299 @@
+    ///       Since this function posts a task which will wait on all others, getStatistics().getPostedTaskCount()
+    ///       will contain one extra count.
+    bool drain(std::chrono::milliseconds timeout = std::chrono::milliseconds(-1),
+               bool isFinal = false);
+
+private:
+    using PendingTaskQueueMap = std::unordered_map<SequenceKey, SequencerLiteKeyData<SequenceKey>, Hash, KeyEqual, Allocator>;
+    using ExceptionCallback = typename Configuration::ExceptionCallback;
+
+    /// @brief Adds the task to the pending queue
+    /// @param key the key associated with the queue to add to
+    /// @param task the task to add
+    /// @return true if the added task is the only enqueued task
+    bool addPendingTask(const SequenceKey& key,
+                        const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task);
+
+    /// @brief Adds the task to the universal pending queue
+    /// @param task the task to add
+    /// @return true if the added task is the only enqueued task
+    bool addPendingTask(const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task);
+
+    /// @brief Schedules the task via the dispatcher
+    /// @param task the task to schedule
+    void scheduleTask(
+        const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task);
+
+    /// @brief Removes the task from the pending queues and schedule next tasks
+    /// @param task the task to remove
+    void removePending(
+        const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task);
+
+    /// @brief Removes the task from the pending queue
+    /// @param entry the queue container
+    /// @param task the task to remove
+    /// @return next task to be scheduled
+    std::shared_ptr<SequencerLiteTask<SequenceKey>> removePending(
+        SequencerLiteKeyData<SequenceKey>& entry,
+        const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task);
+
+    template <class FUNC, class ... ARGS>
+    void
+    enqueueSingle(void* opaque, int queueId, bool isHighPriority, const SequenceKey& sequenceKey, FUNC&& func, ARGS&&... args);
+
+    template <class FUNC, class ... ARGS>
+    void
+    enqueueMultiple(void* opaque,
+                    int queueId,
+                    bool isHighPriority,
+                    const std::vector<SequenceKey>& sequenceKeys,
+                    FUNC&& func,
+                    ARGS&&... args);
+
+    template <class FUNC, class ... ARGS>
+    void
+    enqueueAllImpl(void* opaque, int queueId, bool isHighPriority, FUNC&& func, ARGS&&... args);
+
+    template<class FUNC, class ...ARGS>
+    static std::function<int(VoidContextPtr)> wrap(FUNC&& func, ARGS&&... args);
+
+    Dispatcher&                  _dispatcher;
+    std::atomic_bool             _drain;
+    SequencerLiteKeyData<SequenceKey> _universalTaskQueue;
+    PendingTaskQueueMap          _pendingTaskQueueMap;
+    ExceptionCallback            _exceptionCallback;
+    std::mutex                   _mutex;
```

Comment:
For consistency I'd suggest we use the `quantum::Mutex` here and its associated guards in the implementation.

Reply:
using quantum::mutex requires to have a pointer to quantum::context. We don't have an active context in SeqeunecerLite::enqueue.

Reply:
We should be able to use `quantum::local::context()` here; if I'm recalling correctly that will work correctly in a coroutine or out of one.

Reply:
quantum::local::context() will return nullptr outside of a coroutine. quantum::local is for accessing coroutine local storage (similar to TLS (thread local storage) for threads).

Reply:
Right, and if we try to lock the `quantum::Mutex` from outside a coroutine, calling `lock(nullptr)` is the same as calling `lock()`; that's what we want, and works regardless of whether we are in a coroutine or not.